Source/WebCore/animation/AnimationTimelinesController.cpp

```diff
@@ -310,75 +311,76 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+    }).iterator->value;
 }
 
-ScrollTimeline* AnimationTimelinesController::scrollTimelineForName(const AtomString& name) const
+void AnimationTimelinesController::registerNamedScrollTimeline(const AtomString& name, const Element& source, ScrollAxis axis)
 {
-    return m_nameToScrollTimelineMap.get(name);
+    auto& timelines = timelinesForName(name);
+
+    auto existingTimelineIndex = timelines.findIf([&](auto& timeline) {
+        return is<ScrollTimeline>(timeline) && originatingElement(timeline) == &source;
+    });
+
+    if (existingTimelineIndex != notFound) {
+        auto& existingScrollTimeline = downcast<ScrollTimeline>(timelines[existingTimelineIndex].get());
+        existingScrollTimeline.setAxis(axis);
+    } else {
+        auto newScrollTimeline = ScrollTimeline::create(name, axis);
+        newScrollTimeline->setSource(&source);
+        timelines.append(WTFMove(newScrollTimeline));
+    }
 }
 
-void AnimationTimelinesController::registerNamedViewTimeline(const AtomString& name, Element& subject, ScrollAxis axis, ViewTimelineInsets&& insets)
+void AnimationTimelinesController::registerNamedViewTimeline(const AtomString& name, const Element& subject, ScrollAxis axis, ViewTimelineInsets&& insets)
 {
-    auto& timelines = m_nameToViewTimelinesMap.ensure(name, [] {
-        return Vector<Ref<ViewTimeline>> { };
-    }).iterator->value;
+    auto& timelines = timelinesForName(name);
 
     auto existingTimelineIndex = timelines.findIf([&](auto& timeline) {
-        return timeline->subject() == &subject;
+        return is<ViewTimeline>(timeline) && originatingElement(timeline) == &subject;
     });
 
     if (existingTimelineIndex != notFound) {
-        auto& existingViewTimeline = timelines[existingTimelineIndex];
-        existingViewTimeline->setAxis(axis);
-        existingViewTimeline->setInsets(WTFMove(insets));
+        auto& existingViewTimeline = downcast<ViewTimeline>(timelines[existingTimelineIndex].get());
+        existingViewTimeline.setAxis(axis);
+        existingViewTimeline.setInsets(WTFMove(insets));
     } else {
         auto newViewTimeline = ViewTimeline::create(name, axis, WTFMove(insets));
         newViewTimeline->setSubject(&subject);
         timelines.append(WTFMove(newViewTimeline));
     }
 }
 
-void AnimationTimelinesController::unregisterNamedViewTimelineForSubject(const AtomString& name, const Element& subject)
+void AnimationTimelinesController::unregisterNamedTimeline(const AtomString& name, const Element& element)
 {
-    auto it = m_nameToViewTimelinesMap.find(name);
-    if (it == m_nameToViewTimelinesMap.end())
-        return;
-
-    auto& timelines = it->value;
-    timelines.removeFirstMatching([&](auto& timeline) {
-        return timeline->subject() == &subject;
-    });
-
-    if (timelines.isEmpty())
-        m_nameToViewTimelinesMap.remove(it);
+    auto it = m_nameToTimelineMap.find(name);
+    if (it != m_nameToTimelineMap.end()) {
```

Comment:
I prefer the early return approach as used above.

Reply:
Fixed.